Add support for Nuvoton numaker m46x SoC series

cyliangtw opened the pull request:
This PR is to add Nuvoton NuMaker M46x SoC series with board NuMaker-PFM-M467.

Depends on https://github.com/zephyrproject-rtos/hal_nuvoton/pull/5

Currently this hal module was approved and is waiting for this Zephyr side PR.

Support soc init. and basic drivers to fulfill hello_world & blinky samples.
Basic driver list as below:


- pinctrl, clock-control, uart/serial, gpio

cyliangtw committed:
soc: arm: add support for nuvoton numaker m46x series

Add initial support for nuvoton numaker m46x SoC series including
basic init.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

cyliangtw committed:
drivers: pinctrl: add support for NuMaker series pinctrl

Add Nuvoton numaker series pinctrl support.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

cyliangtw committed:
drivers: clock_control: add support for Nuvoton numaker series CLK

Add Nuvoton numaker series clock controller support, including:
1.  Do system clock initialization in z_arm_platform_init().
2.  Support peripheral clock control API equivalent to BSP
    CLK_EnableModuleClock()/CLK_SetModuleClock().

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

cyliangtw commented:
@stephanosio , the rootcause of failure on `"Run tests with twister / twister-build (4) & build (5)"` is `"fatal error: m46x/sys.h: No such file or directory"`. It could pass these CI tests depending on https://github.com/zephyrproject-rtos/hal_nuvoton/pull/5.

For reference, attach my local build/test log files of syscalls & zexpect as below:
[syscalls_test_log.txt](https://github.com/zephyrproject-rtos/zephyr/files/10893885/syscalls_test_log.txt)
[zexpect_build_log.txt](https://github.com/zephyrproject-rtos/zephyr/files/10893886/zexpect_build_log.txt)
[zexpect_test_log.txt](https://github.com/zephyrproject-rtos/zephyr/files/10893887/zexpect_test_log.txt)
[syscalls_build_log.txt](https://github.com/zephyrproject-rtos/zephyr/files/10893888/syscalls_build_log.txt)

henrikbrixandersen commented:
> It could pass these CI tests depending on [zephyrproject-rtos/hal_nuvoton#5](https://github.com/zephyrproject-rtos/hal_nuvoton/pull/5).

You need to add that change to west.yaml (`revision: pull/5/head`) in this PR following the instructions in https://docs.zephyrproject.org/latest/develop/modules.html#submitting-new-modules

dcpleung commented on drivers/serial/uart_numaker.c:
```diff
@@ -0,0 +1,484 @@
+
+	/* Set pinctrl for UART0 RXD and TXD */
+	/* Set multi-function pins for UART0 RXD and TXD */
+#ifdef CONFIG_PINCTRL
+	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (err != 0) {
+		goto move_exit;
+	}
+#else
+	/* For M46x EVB debug port */
+	SET_UART0_RXD_PB12();
+	SET_UART0_TXD_PB13();
+#endif
```
Could the `#else` part be done via board config in dts? You can mandate that this driver requires `PINCTRL` so it can always set the pins. I am worried about the time when developers would disable `PINCTRL` manually for debugging and the else part is executed.

cyliangtw replied:
Thanks of your consideration.
The `#else` part is one specific case and a little bit confused.
So, I removed it in commit [7295a30](https://github.com/zephyrproject-rtos/zephyr/pull/55406/commits/7295a308aaa8ce08cbb207a4d2ae7234503f88b6).

cyliangtw replied:
@dcpleung , 
Hi Daniel,
I rebase to fix CI failure of `"\tests\bsim folder not foud"`.
So, the relative commit changed as [91b193c](https://github.com/zephyrproject-rtos/zephyr/pull/55406/commits/91b193cbc3d373b16a83257de0483536d6bf1dce)

stephanosio commented:
@cyliangtw https://github.com/zephyrproject-rtos/hal_nuvoton/pull/5 has been merged.

cyliangtw commented:
Rebase to fix `Bluetooth Tests / bluetooth-test (pull_request)` failure caused by \tests\bsim folder not foud.

dcpleung reviewed (dismissed):
UART looks good.

cyliangtw commented:
> UART looks good.

@dcpleung , hi Daniel,

Yesterday, 88cedcf5c588 was merged and my branch got `Conflicting files clock_control/CMakeLists.txt & Kconfig`.

Should I process rebase & fix it now or wait until pass the necessary reviewers ?

dcpleung commented:
It would be good to resolve any conflicts so others are reviewing merge-able code.

cyliangtw commented:
> It would be good to resolve any conflicts so others are reviewing merge-able code.

@dcpleung ,  hi Daniel,
I rebase to fix clock_control/CMakeLists.txt conflict.
Now, it's merge-able but your approval seems dismissed.

cyliangtw commented:
Hi @MulinChao , @ChiHuaL  please have a look into this PR

cyliangtw commented:
Hi @henrikbrixandersen , @stephanosio,
Rebse to fix serial/Kconfig conflicts, now it's merge-able.
Please have a look into this PR.

MulinChao commented on drivers/gpio/gpio_numaker.c:
```diff
@@ -0,0 +1,287 @@
+		(config->reg - DT_REG_ADDR(DT_NODELABEL(gpioa))) / DT_REG_SIZE(DT_NODELABEL(gpioa));
+	uint32_t *GPx_MFPx = ((uint32_t *)&SYS->GPA_MFP0) + port_index * 4 + (pin / 4);
+	/* E.g.: SYS->GPA_MFP0  = (SYS->GPA_MFP0 & (~SYS_GPA_MFP0_PA0MFP_Msk) ) | */
+	/* SYS_GPA_MFP0_PA0MFP_SC0_CD  ; */
```
Redundant space and ditto for multi-lines commenting.

cyliangtw replied:
Fixed in commit d2bd74bc5b.

MulinChao reviewed (commented):
LGTM. Few comments for code style.

cyliangtw commented on drivers/gpio/gpio_numaker.c:
```diff
@@ -0,0 +1,287 @@
+
+	/* Set Multi-function, default is GPIO */
+	uint32_t port_index =
+		(config->reg - DT_REG_ADDR(DT_NODELABEL(gpioa))) / DT_REG_SIZE(DT_NODELABEL(gpioa));
```
Yes, it's better to use config->gpa_base. Fixed it in commit d2bd74bc5bf27.

cyliangtw commented on boards/arm/numaker_pfm_m467/Kconfig.board:
```diff
@@ -0,0 +1,9 @@
+#
+# Nuvoton PFM M467 board configuration
+#
+# Copyright (c) 2022 Nuvoton Technology Corporation.
```
Developed in 2022/12.
Now, correct relative files as Copyright (c) 2023.

cyliangtw commented on boards/arm/numaker_pfm_m467/numaker_pfm_m467-pinctrl.dtsi:
```diff
@@ -0,0 +1,52 @@
+				 <PB13MFP_UART0_TXD 0x0000>;
+	};
+
+/* TX/RX/RTS/CTS/RST --> D1/D0/A2/A3/D2 --> PB3/PB2/PB8/PB9/PC9 */
```
Add one Tab to keep consistent format, fixed in commit  e3b62d97.

cyliangtw commented on drivers/clock_control/clock_control_numaker_scc.c:
```diff
@@ -0,0 +1,160 @@
+	LOG_INF("Core clock: %d (Hz)", cfg->core_clock);
+#endif
+
+	/* z_arm_platform_init() will respect above configurations and  */
```
Thanks of your advice.
Modified all files with multi-lines comments, to follow the preferred style.

cyliangtw commented on drivers/serial/uart_numaker.c:
```diff
@@ -0,0 +1,480 @@
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+
```
Yes, removed redundant line in commit  77b5746.

fabiobaltieri commented on soc/arm/nuvoton_numaker/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,33 @@
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/types.h>
+
+#define PORT_INDEX(pinmux)    (((pinmux)&0xF0000000) >> 28)
```
You could use `FIELD_GET` and `GENMASK` for this pattern.

cyliangtw replied:
Thanks of your remind. I will consider which one could be  easier mapping v.s data sheet.

fabiobaltieri commented on drivers/clock_control/clock_control_numaker_scc.c:
```diff
@@ -0,0 +1,162 @@
+		CLK_EnableModuleClock(scc_subsys->pcc.clk_modidx);
+		SYS_LockReg();
+	} else {
+		return -EINVAL;
```
I'd suggest inverting the logic here, change to

```
if (scc_subsys->subsys_id != NUMAKER_SCC_SUBSYS_ID_PCC) {
        return -EINVAL;
}
```

and let the normal execution flow with no indentation. Same below, no big deal in this case but generally makes code easier to follow.

cyliangtw replied:
This code structure is reserved to support future subsys_id definitions if need be.
for ex: NUMAKER_SCC_SUBSYS_ID_PCC2, NUMAKER_SCC_SUBSYS_ID_PCC3

fabiobaltieri commented on include/zephyr/drivers/clock_control/clock_control_numaker.h:
```diff
@@ -0,0 +1,41 @@
+	 * clk_src is same as u32ClkSrc in BSP CLK_SetModuleClock().
+	 * clk_div is same as u32ClkDiv in BSP CLK_SetModuleClock().
+	 */
+	union {
```
Do you still need the union?

cyliangtw replied:
Ditto.

fabiobaltieri commented on drivers/serial/uart_numaker.c:
```diff
@@ -0,0 +1,479 @@
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+/* API implementation: fifo_fill */
```
I would drop these comments, they are really just repeating the same information that you have in the function name itself.

cyliangtw replied:
Agree with you, it's redundant.

fabiobaltieri commented on drivers/gpio/gpio_numaker.c:
```diff
@@ -0,0 +1,289 @@
+
+	/* Set Multi-function, default is GPIO */
+	uint32_t port_index =
+	(config->reg - config->gpa_base) / config->size;
```
odd indentation: bring on the previous line

cyliangtw replied:
Previous version with macro of get-base & get-size, so got WARNING:LONG_LINE at that time.
Now, it could be under limit of long line. 
Thanks of your remind.

fabiobaltieri commented on drivers/gpio/gpio_numaker.c:
```diff
@@ -0,0 +1,289 @@
+	 * E.g.: SYS->GPA_MFP0  = (SYS->GPA_MFP0 & (~SYS_GPA_MFP0_PA0MFP_Msk) ) |
+	 * SYS_GPA_MFP0_PA0MFP_SC0_CD;
+	 */
+	*GPx_MFPx = (*GPx_MFPx & (~pinMfpMask)) | 0x00;
```
Uh... what does the `| 0x00` do here?

cyliangtw replied:
It's the multi-function value of gpio group.
To modify as readable: `*GPx_MFPx = (*GPx_MFPx & (~pinMfpMask)) | pinMfpGpio;` 
 in commit d9bfaad3f.

fabiobaltieri commented on drivers/gpio/gpio_numaker.c:
```diff
@@ -0,0 +1,289 @@
+#define GPIO_NUMAKER_DEFINE(n)                                                                     \
+	static const struct gpio_numaker_config gpio_numaker_config##n = {                         \
+		.common =                                                                          \
+			{                                                                          \
```
Bring the `{` on the previous line.

cyliangtw replied:
Appreciate & fix in commit d9bfaad3f.

fabiobaltieri commented on dts/arm/nuvoton/m46x.dtsi:
```diff
@@ -0,0 +1,586 @@
+				interrupt-names = "LINE_0", "LINE_1";
+				reset = <CANFD0_RST>;
+				clocks = <&pcc CANFD0_MODULE CLK_CLKSEL0_CANFD0SEL_HCLK \
+CLK_CLKDIV5_CANFD0(1)>;
```
Oh ok, yeah if the CI lets you do it I'd say it's fine, otherwise if you want to indent it that's fine as well but align it with the `&`, something like

````
clocks = <&pcc
          CANFD0_MODULE CLK_CLKSEL0_CANFD0SEL_HCL 
          CLK_CLKDIV5_CANFD0(1)>;
```

Don't think the `\` is needed.

cyliangtw commented on boards/arm/numaker_pfm_m467/doc/pfm_m467.jpeg:
yes, it could save 18%.

cyliangtw commented on boards/arm/numaker_pfm_m467/numaker_pfm_m467-pinctrl.dtsi:
```diff
@@ -0,0 +1,52 @@
+	/* TX/RX/RTS/CTS/RST --> D1/D0/A2/A3/D2 --> PB3/PB2/PB8/PB9/PC9 */
+	uart1_default: uart1_default {
+		pinmux = <PB9MFP_UART1_nCTS 0x0000>,
+				<PB8MFP_UART1_nRTS 0x0000>,
```
it's fine to realign.

henrikbrixandersen commented on dts/arm/nuvoton/m46x.dtsi:
```diff
@@ -0,0 +1,590 @@
+			rx-buffer-elements = <3>;
+			tx-buffer-elements = <3>;
+
+			canfd0: canfd@40020000 {
+				compatible = "nuvoton,numaker";
+				reg = <0x40020000 0x200>, <0x40020200 0x1800>;
+				reg-names = "m_can", "message_ram";
+				interrupts = <112 0>, <113 0>;
+				interrupt-names = "LINE_0", "LINE_1";
+				reset = <CANFD0_RST>;
+				clocks = <&pcc
+						  CANFD0_MODULE CLK_CLKSEL0_CANFD0SEL_HCLK
+						  CLK_CLKDIV5_CANFD0(1)>;
+				status = "disabled";
+				sjw = <1>;
+				sample-point = <875>;
+				sjw-data = <1>;
+				sample-point-data = <875>;
+			};
```
Please delay adding CAN nodes until after a corresponding DTS binding/driver support has been added. Same goes for remaining DTS nodes without corresponding bindings and/or drivers.

cyliangtw replied:
Hi @henrikbrixandersen ,
In fact, corresponding driver of DTS nodes are finished and tested.
To fulfill small PR, so split & only keep basic drivers in 1st PR of m46x.
Then, plan to send several PRs for remainding drivers one by one.
For this situation, should I remove CAN/spi/pwm/usb nodes ?

henrikbrixandersen replied:
No. If corresponding drivers are ready for submission after this initial PR, I think you can keep them.

The danger here is that a review of those follow-up PRs will result in changes to those bindings and thus a need to revisit the DTSI.

cyliangtw replied:
Thanks of your advice, corresponding drivers are ready for submission after this PR.
I appreciate your remind.

gmarull commented on dts/arm/nuvoton/m46x.dtsi:
```diff
@@ -0,0 +1,590 @@
+			compatible = "nuvoton,numaker-uart";
+			reg = <0x40070000 0x1000>;
+			interrupts = <36 0>;
+			reset = <UART0_RST>;
```
where is the reset controller here?

cyliangtw replied:
In `uart_numaker_init()`, call `SYS_ResetModule(config->id_rst)` to ensure H/W status clean.

gmarull replied:
this needs a reset driver

gmarull replied:
ST/GD and other platforms are as simple as that, but for consistency we should be using reset controller API.

cyliangtw replied:
I see, most MPU chip with asert/deassert feature will need Linux reset driver.
I will check the capibility with only toggle function, how to fulfill.

cyliangtw replied:
After check, it's doable and is great to consistency.
For simplified this PR, could I add reset driver in next PR ?

cyliangtw replied:
Hi @gmarull,
Sorry for late reply due to consecutive holidays, 
I added the reset driver at commit [c92783d](https://github.com/zephyrproject-rtos/zephyr/pull/55406/commits/c92783ddb9b3f91afcdebf594b92209dab83dde0) in this PR.

cyliangtw committed:
drivers: reset: add support for NuMaker series reset

Add Nuvoton numaker series reset controller support.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

cyliangtw committed:
drivers: serial: support for Nuvoton numaker series UART

Add Nuvoton numaker series UART support, including interrupt-driven,
also apply pinctrl and clock-control.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

cyliangtw committed:
drivers: gpio: support for Nuvoton numaker series GPIO

Add Nuvoton numaker series GPIO support, including interrupt mode and
also integrate clock control.

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

cyliangtw commented:
Hi  @gmarull , @dcpleung , @MulinChao , @MaureenHelm,

Add the reset driver at commit [c92783d](https://github.com/zephyrproject-rtos/zephyr/pull/55406/commits/c92783ddb9b3f91afcdebf594b92209dab83dde0).
Also apply `reset_line_toggle()` to reset uart H/W IP in serial driver commit [bebcbff](https://github.com/zephyrproject-rtos/zephyr/pull/55406/commits/bebcbff6a5353fbd4037a121f4f7f8f8a8e6f68b).

Please have a look into this change.

gmarull commented on drivers/pinctrl/pinctrl_numaker.c:
```diff
@@ -0,0 +1,56 @@
+#include <soc.h>
+
+struct nvt_pinctrl_pin {
+	uint32_t pinMux;
+	uint32_t cfgData;
```
please don't use camelCase

gmarull commented on dts/bindings/pinctrl/nuvoton,numaker-pinctrl.yaml:
```diff
@@ -0,0 +1,78 @@
+        - "low"
+        - "fast"
+      description: |
+        Set the driving strength of a pin.
```
defaults need justification

cyliangtw replied:
fixed in 3f90ae5816

gmarull commented on dts/bindings/pinctrl/nuvoton,numaker-pinctrl.yaml:
```diff
@@ -0,0 +1,78 @@
+        - "fast"
+        - "low"
+      description: |
+        Set the speed of a pin. This setting effectively limits the
+        slew rate of the output signal.
```
ditto for default

cyliangtw replied:
Done

gmarull commented on drivers/clock_control/clock_control_numaker_scc.c:
```diff
@@ -0,0 +1,162 @@
+
+#define DT_DRV_COMPAT nuvoton_numaker_scc
+
+#include <soc.h>
```
avoid using soc.h, but HAL header directly.

cyliangtw replied:
Done

gmarull commented on dts/bindings/clock/nuvoton,numaker-scc.yaml:
```diff
@@ -0,0 +1,49 @@
+  reg:
+    required: true
+
+  hxt:
```
why not make this a boolean,  eg `hxt-enable`? Same for others.

cyliangtw replied:
Thanks of your advice. Reserve for one more choice "untouched".

gmarull commented on drivers/serial/uart_numaker.c:
```diff
@@ -0,0 +1,478 @@
+#define CLOCK_CTRL_INIT(n)
+#endif
+
+#ifdef CONFIG_PINCTRL
+#define PINCTRL_DEFINE(n) PINCTRL_DT_INST_DEFINE(n);
+#define PINCTRL_INIT(n)	  .pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),
+#else
+#define PINCTRL_DEFINE(n)
+#define PINCTRL_INIT(n)
+#endif
```
please don't make this conditional, pinctrl is the only way now to configure pins (we need to do some cleanup on this)

cyliangtw replied:
OK, regard pinctrl as the only way.

gmarull commented on drivers/gpio/Kconfig.numaker:
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config GPIO_NUMAKER
+	bool "Nuvoton NUMAKER MCU gpio driver"
+	select HAS_NUMAKER_GPIO
+	depends on SOC_FAMILY_NUMAKER
```
DT dependency missing?

cyliangtw replied:
Add DT_COMPAT_NUVOTON_NUMAKER_GPIO

gmarull commented on drivers/gpio/gpio_numaker.c:
```diff
@@ -0,0 +1,291 @@
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <soc.h>
```
not needed, you're including HAL below. Please, sort headers better.

gmarull commented on drivers/gpio/gpio_numaker.c:
```diff
@@ -0,0 +1,291 @@
+		return -EINVAL;
+	}
+
+	SYS_UnlockReg();
```
can you please detail why we need all these SYS_* calls?

cyliangtw replied:
In our specific MCU H/W design, we need to unlock H/W before some manipulations. Our MPU chip no this kind control.

cyliangtw commented on soc/arm/nuvoton_numaker/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,33 @@
+
+#include <zephyr/devicetree.h>
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/types.h>
```
Fixed in ea2a079bc2

cyliangtw commented on soc/arm/nuvoton_numaker/m46x/soc.c:
```diff
@@ -0,0 +1,87 @@
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/clock_control/clock_control_numaker.h>
+/* Hardware and starter kit includes. */
+#include "NuMicro.h"
```
Fixed in https://github.com/zephyrproject-rtos/zephyr/commit/ea2a079bc2b3dded194a9569a2a2cb61eb32f294

cyliangtw commented on soc/arm/nuvoton_numaker/m46x/soc.h:
```diff
@@ -0,0 +1,14 @@
+#ifndef ZEPHYR_SOC_ARM_NUVOTON_M46X_SOC_H_
+#define ZEPHYR_SOC_ARM_NUVOTON_M46X_SOC_H_
+
+#include <zephyr/sys/util.h>
```
Done

gmarull commented on dts/bindings/pinctrl/nuvoton,numaker-pinctrl.yaml:
```diff
@@ -0,0 +1,80 @@
+        - "fast"
+        - "low"
+      description: |
+        Set the speed of a pin. This setting effectively limits the
+        slew rate of the output signal. Default to fast for the capability to
+        support high speed pins, like as SPI CLK up to 50MHz.
```
What is hardware default? `fast` default is, in general, not desirable as it can cause EMI issues if the circuit is not designed for high speed in mind.

cyliangtw replied:
Done

cyliangtw commented on dts/bindings/pinctrl/nuvoton,numaker-pinctrl.yaml:
```diff
@@ -0,0 +1,80 @@
+        - "low"
+        - "fast"
+      description: |
+        Set the driving strength of a pin. Default to low is enough to drive most
+        components, like as LED, CAN transceiver and son on.
```
Yes, default is low in data sheet. 
So, should I revise as "Default to low in hardware register" ?

cyliangtw replied:
Thanks of your remind and revised as:
```
Hardware default configuration is low and
it's enough to drive most components, like as LED, CAN transceiver and son on.
```

gmarull commented on dts/bindings/pinctrl/nuvoton,numaker-pinctrl.yaml:
```diff
@@ -0,0 +1,80 @@
+  properties:
+    pinmux:
+      required: true
+      type: array
```
why is this an array? I see second field is always `0`

cyliangtw replied:
For example uart/serial port, it may need TX/RX/CTS/RTS, so it's an array.
like as:
```
	uart1_default: uart1_default {
		pinmux = <PB9MFP_UART1_nCTS 0x0000>,
			 <PB8MFP_UART1_nRTS 0x0000>,
			 <PB2MFP_UART1_RXD 0x0000>,
			 <PB3MFP_UART1_TXD 0x0000>,
```

gmarull replied:
sorry, I meant the `0x0000`

cyliangtw replied:
it's the reserved config data for some complex NuMaker chip.
For M46x, 1st field 'PB9MFP_UART1_nCTS' is enough to encode all pinmux information.
For some complex NuMaker chip, one field can't encode all information.
Sorry, it's my consideration for compatible all NuMaker chips in near future.

gmarull replied:
Note: fields are 32-bit maximum, so you can just let e.g. `PB9MFP_UART1_nCTS` use 32-bit, with unused 16-bit. This way you don't need this 16-bit+16-bit(all-zero) pattern, it looks weird. Can you please justify a bit this "future requirements"? Why a single 32-bit field is not an option?

cyliangtw replied:
Sorry, 0x0000 is a little confused, in fact, both of 1st field (i.e PB9MFP_UART1_nCTS) and 2nd field are 32-bit.  It will be better as <PB9MFP_UART1_nCTS 0x00000000>.
Current Bit[31-28] for pin group, Bit[27-24] for pins of one group. As while groups over 16 or pins of one group over 16, it will need 2nd field.

cyliangtw replied:
After discussion with H/W architecture designer, it's feasible to limit MCU multi-function pins under 16*16 criteria in the future design. So, I will remove 2nd field 0x0000.

cyliangtw replied:
Thanks of your review.
Finished the modification of pin-mux with a single 32-bit field.

cyliangtw replied:
Hi @gmarull ,
Done and force pushed branch for this modification of pin-mux.

cyliangtw replied:
Hi @gmarull ,
  Please have a look into this update.

gmarull commented on dts/bindings/pinctrl/nuvoton,numaker-pinctrl.yaml:
```diff
@@ -0,0 +1,80 @@
+        - "fast"
+      description: |
+        Set the driving strength of a pin. Hardware default configuration is low and
+        it's enough to drive most components, like as LED, CAN transceiver and son on.
```
`so on`

cyliangtw replied:
To remove "so on", just keep "like as LED and CAN transceiver."
Right ?

gmarull replied:
`s/son on/so on`

cyliangtw replied:
I see, typo issue

cyliangtw replied:
Done.

cyliangtw commented on soc/arm/nuvoton_numaker/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,32 @@
+/*
```
OK, I will move to  pinctrl commit.

cyliangtw replied:
it's done.

cyliangtw commented:
Rebase and fixed warning of "Run compliance checks on patch series".

gmarull commented on drivers/pinctrl/Kconfig.numaker:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2022 Nuvoton Technology Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_NUVOTON_NUMAKER_PINCTRL := nuvoton,numaker-pinctrl
+
+config PINCTRL_NUMAKER
+	bool "Nuvoton NuMaker pin controller driver"
+	depends on SOC_FAMILY_NUMAKER
+	default $(dt_compat_enabled,$(DT_COMPAT_NUVOTON_NUMAKER_PINCTRL))
+	help
```
please use new `DT_` predefined Kconfig options, check other drivers

cyliangtw replied:
Done.

gmarull commented on drivers/pinctrl/pinctrl_numaker.c:
```diff
@@ -0,0 +1,47 @@
+
+#define DT_DRV_COMPAT nuvoton_numaker_pinctrl
+
+#include <assert.h>
```
not used

cyliangtw replied:
OK, I remove it.

gmarull commented on soc/arm/nuvoton_numaker/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,32 @@
+#define NU_MFP_MASK(pinindex) (0x1f << NU_MFP_POS(pinindex))
+
+typedef struct pinctrl_soc_pin_t {
+	uint32_t pinItem;
```
nit: no camelCase

gmarull commented on drivers/clock_control/clock_control_numaker_scc.c:
```diff
@@ -0,0 +1,162 @@
+
+#define LOG_OSC_SW(osc, sw)                                                                        \
+	if (sw == NUMAKER_SCC_CLKSW_ENABLE) {                                                      \
+		LOG_INF("Enable " #osc);                                                           \
+	} else if (sw == NUMAKER_SCC_CLKSW_DISABLE) {                                              \
+		LOG_INF("Disable " #osc);                                                          \
```
LOG_DBG?

cyliangtw replied:
Yes, LOG_DBG is better than LOG_INF.

cyliangtw replied:
done

gmarull commented on drivers/clock_control/clock_control_numaker_scc.c:
```diff
@@ -0,0 +1,162 @@
+{
+	const struct numaker_scc_config *cfg = dev->config;
+
+	LOG_INF("CLK base: 0x%08x", cfg->clk_base);
```
ditto

cyliangtw replied:
done

gmarull commented on drivers/reset/Kconfig.numaker:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2023 Nuvoton Technology Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_NUVOTON_NUMAKER_RST := nuvoton,numaker-rst
+
+config RESET_NUMAKER
+	bool "Nuvoton NuMaker reset controller driver"
+	depends on SOC_FAMILY_NUMAKER
+	default $(dt_compat_enabled,$(DT_COMPAT_NUVOTON_NUMAKER_RST))
+	help
+	  This option enables the reset controller driver for Nuvoton NuMaker MCUs.
```
ditto

cyliangtw replied:
done

gmarull commented on drivers/serial/Kconfig.numaker:
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2020 Nuvoton Technology Corporation.
+# SPDX-License-Identifier: Apache-2.0
+
+config UART_NUMAKER
+	bool "Nuvoton NUMAKER MCU serial driver"
+	select SERIAL_HAS_DRIVER
+	select HAS_NUMAKER_UART
+	select SERIAL_SUPPORT_INTERRUPT
+	depends on SOC_FAMILY_NUMAKER
```
missing DT dependency

cyliangtw replied:
Done

gmarull commented on dts/bindings/serial/nuvoton,numaker-uart.yaml:
```diff
@@ -0,0 +1,24 @@
+  clocks:
+    required: true
+
+  pinctrl-0:
```
require pinctrl-names as well?

cyliangtw replied:
Thanks of this review.
Above issues are fixed and updated.

cyliangtw replied:
Hi @gmarull,
Fixed and please have a look into the update.

gmarull commented on boards/arm/numaker_pfm_m467/CMakeLists.txt:
```diff
@@ -0,0 +1 @@
+# SPDX-License-Identifier: Apache-2.0
```
empty cmake? can you also add board after adding SoC support to preserve bisectability?

cyliangtw replied:
OK, it's fine to remove this empty CMakeLists.txt .
About preserve bisectability, is it meaning to exchange the commit sequence of "board" and "soc", to let soc as 1st commit ?

gmarull replied:
yes, board should be the latest.

gmarull commented on boards/arm/numaker_pfm_m467/numaker_pfm_m467_defconfig:
```diff
@@ -0,0 +1,25 @@
+CONFIG_BOARD_NUMAKER_PFM_M467=y
+CONFIG_PINCTRL=y
+CONFIG_GPIO=y
+CONFIG_GPIO_NUMAKER=y
```
redundant

gmarull commented on boards/arm/numaker_pfm_m467/numaker_pfm_m467_defconfig:
```diff
@@ -0,0 +1,25 @@
+
+# enable uart driver
+CONFIG_SERIAL=y
+CONFIG_UART_NUMAKER=y
```
redundant

cyliangtw replied:
It's OK to remove CONFIG_UART_NUMAKER=y.
But need keep CONFIG_SERIAL=y, else hello_world sample can't link uart_numaker.o .
Is it correct ?

gmarull replied:
Yes, only driver class is needed, other is automatically selected by DT status.

gmarull commented on soc/arm/nuvoton_numaker/Kconfig.defconfig:
```diff
@@ -0,0 +1,8 @@
+
+source "soc/arm/nuvoton_numaker/*/Kconfig.defconfig.series"
+
+config RESET
+	default y
```
there's no reset driver at this point, please, add stuff in a bisectable way. in this case, this needs to be added by the introduction of the reset driver.

cyliangtw replied:
So, should I move this update of Kconfig.defconfig in the commit of reset driver ?

gmarull replied:
yes please, also review for any other similar issues.

cyliangtw replied:
Thanks, I also move the clock control update of soc.c to the commit of clock-control driver.
Fixed above issues and force-pushed commits.

cyliangtw replied:
Hi @gmarull,
Fixed and please have a look.

cyliangtw replied:
Hi @gmarull,
I updated the code based on your requested changes.
Please have a look.

cyliangtw replied:
Hi @gmarull,
Finished your requested changes.
Please revisit.

cyliangtw commented:
Rebase and force-pushed branch to resolve error of `Pull Request Assigner / Pull Request Assignment (pull_request_target)`

cfriedt commented:
It's a bit confusing why we have Zephyr drivers like clock control but then also hal alternatives in case the Zephyr driver isn't enabled... can't we just have GPIO depend on CLOCK_CONTROL?

cyliangtw commented:
> It's a bit confusing why we have Zephyr drivers like clock control but then also hal alternatives in case the Zephyr driver isn't enabled... can't we just have GPIO depend on CLOCK_CONTROL?

In the developing scenario, clock control driver is later than serial & gpio drivers. So, serial & gpio drivers use hal alternative firstly, then apply Zephyr clock control API later.
Yes, it's fine to regard Zephyr clock control API as the only way.

gmarull commented on dts/arm/nuvoton/m46x.dtsi:
```diff
@@ -0,0 +1,597 @@
+			};
+		};
+
+		rst: reset-controller@40000000 {
+			compatible = "nuvoton,numaker-rst";
+			reg = <0x40000000 0x20>;
+			#reset-cells = <1>;
+			status = "okay";
+		};
```
why is this adding a node without a driver? Please, introduce the relevant nodes in the same commit where you introduce a driver/binding for them.

henrikbrixandersen replied:
This is especially important since otherwise changes may be needed to the binding/dts node once the driver is written. It is difficult to predict the exact binding needed before writing the driver.

cyliangtw replied:
Yes, I see, in fact this dtsi was verified by the relevant dirves, the drivers are ready for next PR one by one.
Should I still need to move the node pinctrl/reset/clock/uart/gpio to relevant drive commit one by one ?

gmarull replied:
yes please

cyliangtw replied:
@henrikbrixandersen & @gmarull,
Thanks of your viewpoint.
Finished the upadte to let node matching driver, please check it.

cyliangtw replied:
Hi @henrikbrixandersen & @gmarull,
If you are available, please look at this update.

gmarull commented on soc/arm/nuvoton_numaker/m46x/soc.c:
```diff
@@ -0,0 +1,61 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
```
not needed

cyliangtw replied:
Not need <zephyr/init.h>.
Keep <zephyr/kernel.h> for DT_NODE_HAS_PROP(DT_NODELABEL(xxx), ...)

gmarull replied:
this is DT API, not Kernel.

cyliangtw replied:
#include <zephyr/device.h> also could resolve the above error.
For DT API, should I include <zephyr/device.h> instead of <zephyr/kernel.h> ?

gmarull commented on drivers/clock_control/Kconfig.numaker:
```diff
@@ -0,0 +1,10 @@
+
+config CLOCK_CONTROL_NUMAKER_SCC
+	bool "NuMaker system clock controller driver"
+	depends on SOC_FAMILY_NUMAKER
```
DT dependency on `nuvoton_numaker_scc`?

cyliangtw replied:
Change to depends on DT_HAS_NUVOTON_NUMAKER_SCC_ENABLED

gmarull commented on drivers/reset/reset_numaker.c:
```diff
@@ -0,0 +1,79 @@
+	.line_toggle = reset_numaker_line_toggle,
+};
+
+static int reset_numaker_init(const struct device *dev)
+{
+	return 0;
+}
```
redundant, just pass NULL to device define macro

gmarull commented on drivers/serial/uart_numaker.c:
```diff
@@ -0,0 +1,451 @@
+#include <zephyr/drivers/clock_control/clock_control_numaker.h>
+#include <zephyr/drivers/pinctrl.h>
+
+#define DT_DRV_COMPAT nuvoton_numaker_uart
```
by convention: move to the top

gmarull commented on drivers/serial/uart_numaker.c:
```diff
@@ -0,0 +1,451 @@
+ *  Copyright (c) 2023 Nuvoton Technology Corporation.
+ */
+
+#include <zephyr/drivers/uart.h>
+#include <zephyr/drivers/reset.h>
+#include <string.h>
+#include <zephyr/irq.h>
+#include <zephyr/logging/log.h>
+#include <NuMicro.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/clock_control_numaker.h>
+#include <zephyr/drivers/pinctrl.h>
```
needs better sorting: libc, zephyr, HAL (check others as well)

cyliangtw replied:
Applied to 5 drivers

gmarull commented on modules/Kconfig.nuvoton:
```diff
@@ -12,9 +12,59 @@ menu "Nuvoton drivers"
+		help
+		  Enable Nuvoton Universal asynchronous receiver transmitter HAL
+		  module driver
+	config HAS_NUMAKER_GPIO
+		bool "NuMaker GPIO"
+		help
+		  Enable Nuvoton gpio HAL module driver
+	config HAS_NUMAKER_FMC
+		bool "NuMaker FMC"
+		help
+		  Enable Nuvoton FMC HAL module driver
+	config HAS_NUMAKER_I2C
+		bool "NuMaker I2C"
+		help
+		  Enable Nuvoton I2C HAL module driver
+	config HAS_NUMAKER_SPI
+		bool "NuMaker SPI"
+		help
+		  Enable Nuvoton SPI HAL module driver
+	config HAS_NUMAKER_PWM
+		bool "NuMaker PWM"
+		help
+		  Enable Nuvoton PWM HAL module driver
+	config HAS_NUMAKER_USBD
+		bool "NuMaker USB 1.1 device controller"
+		help
+		  Enable Nuvoton USB 1.1 device controller HAL module driver
+	config HAS_NUMAKER_HSUSBD
+		bool "NuMaker high-speed USB 2.0 device controller"
+		help
+		  Enable Nuvoton high-speed USB 2.0 device controller HAL module driver
+	config HAS_NUMAKER_ETH
+		bool "NuMaker ETH"
+		help
+		  Enable Nuvoton ETH EMAC HAL module driver
+	config HAS_NUMAKER_CANFD
+		bool "NuMaker CAN FD"
+		help
+		  Enable Nuvoton CAN FD HAL module driver
+endmenu
```
looks like this needs to go into its own commit, before drivers?

gmarull replied:
not solved

gmarull commented on dts/arm/nuvoton/m46x.dtsi:
```diff
@@ -0,0 +1,47 @@
+
+#include <arm/armv7-m.dtsi>
+#include <mem.h>
+#include "m46x/sys.h"
+#include "m46x/sys_reg.h"
+#include "m46x/clk.h"
+#include "m46x/clk_reg.h"
```
then these files need to be part of dt-bindings

cyliangtw replied:
these files are at hal_nuvoton:
https://github.com/zephyrproject-rtos/hal_nuvoton/tree/master/dts/m46x

gmarull replied:
Then please, create proper dt-bindings in Zephyr tree. See for example GD32 reset dt-binding files.

cyliangtw replied:
Hi @gmarull & @henrikbrixandersen ,
  Followed your advice to create proper dt-bindings:
```
#include <zephyr/dt-bindings/pinctrl/numaker_m46x_pinctrl.h>
#include <zephyr/dt-bindings/clock/numaker_m46x_clock.h>
#include <zephyr/dt-bindings/reset/numaker_m46x_reset.h>

```
Please look at it.

dcpleung commented on drivers/serial/uart_numaker.c:
```diff
@@ -0,0 +1,451 @@
+	UART_DisableInt(uart, UART_INTEN_THREIEN_Msk);
+}
+
+#define UART_ENABLE_INT(uart, u32eIntSel) ((uart)->INTEN |= (u32eIntSel))
```
Remove?

cyliangtw replied:
@dcpleung,
Thanks, it's redundancy and I removed it in commit [01d6eec](https://github.com/zephyrproject-rtos/zephyr/pull/55406/commits/01d6eecb045d76b6ed5c0911a9d8c29ea0d9f0a2).

carlescufi commented:
@gmarull please take another look

gmarull commented on include/zephyr/dt-bindings/clock/numaker_m46x_clock.h:
```diff
@@ -0,0 +1,2004 @@
+#define MODULE_CLKDIV_Pos(x) (((x) >> 5) & 0x1fUL)
+#define MODULE_IP_EN_Pos(x)  (((x) >> 0) & 0x1fUL)
+#define MODULE_NoMsk	     0x0UL
+#define NA		     MODULE_NoMsk
```
this file includes > 2k lines, it's not clear to me what is expected to be used in DT. Things like `#define NA` or the lack of proper namespacing look suspicious. Can you please clarify? How is one expected to understand/use this file?

cyliangtw replied:
In DT, it need the definition of clk module, clk select or clk divider, this file include all the options. This chip is for general purpose, so there are many options for each HW IP. Normally, developer will refer to existing property in dtsi  or check TRM doc for the config, just like what developer did on our proprietary BSP.

cyliangtw replied:
Hi @gmarull,
Refined numaker_m46x_clock.h, to keep min. set of the definition of clk-module, clk-select & clk-divider for each H/W IP's clock config.
For ex. `<UART0_MODULE CLK_CLKSEL1_UART0SEL_HIRC CLK_CLKDIV0_UART0(1)>`:

-  UART0_MODULE: clk-module of UART0
-  CLK_CLKSEL1_UART0SEL_HIRC: choose HIRC as UART0's clk source
-  CLK_CLKDIV0_UART0(1): set clk divider for UART0

Please see [00f06a7](https://github.com/zephyrproject-rtos/zephyr/pull/55406/commits/00f06a740da7e685609dfd723e640a699aace213).

gmarull commented on include/zephyr/dt-bindings/pinctrl/numaker_m46x_pinctrl.h:
```diff
@@ -0,0 +1,2154 @@
+/*
```
by convention of all other files, use `numaker-m46x-pinctrl.h` file name

cyliangtw replied:
Yes, it's better to match convention.

gmarull commented on include/zephyr/dt-bindings/pinctrl/numaker_m46x_pinctrl.h:
```diff
@@ -0,0 +1,2154 @@
+ */
+
+/* PA.0 MFP */
+#define SYS_GPA_MFP0_PA0MFP_GPIO         (0x00UL<<SYS_GPA_MFP0_PA0MFP_Pos)
```
please format this file properly, e.g. this is missing spaces around << (try clang-format)

gmarull commented on include/zephyr/dt-bindings/pinctrl/numaker_m46x_pinctrl.h:
```diff
@@ -0,0 +1,2154 @@
+
+/* Beginning of M460 BSP sys_reg.h pin-mux module copy */
+
+#define SYS_GPA_MFP0_PA0MFP_Pos          (0)
+#define SYS_GPA_MFP0_PA1MFP_Pos          (8)
+#define SYS_GPA_MFP0_PA2MFP_Pos          (16)
+#define SYS_GPA_MFP0_PA3MFP_Pos          (24)
```
these need to be namespaced as other pinctrl dt-bindings, e.g. `NUMAKER_M46X_...`

cyliangtw replied:
@gmarull ,
Except xxx_pinctrl.h, xxx_clock.h seems also be requested.
After checking, if add namespace "NUMAKER" for each macro in numaker_m46x_clock.h, the macro will be much longer & not easy read, like as: 
```
#define NUMAKER_UART0_MODULE                                                                               \
	(NUMAKER_MODULE_APBCLK_ENC(1UL) | NUMAKER_MODULE_IP_EN_Pos_ENC(NUMAKER_CLK_APBCLK0_UART0CKEN_Pos) |                \
	 NUMAKER_MODULE_CLKSEL_ENC(1UL) | NUMAKER_MODULE_CLKSEL_Msk_ENC(3UL) | NUMAKER_MODULE_CLKSEL_Pos_ENC(24UL) |       \
	 NUMAKER_MODULE_CLKDIV_ENC(0UL) | NUMAKER_MODULE_CLKDIV_Msk_ENC(0x0FUL) | NUMAKER_MODULE_CLKDIV_Pos_ENC(8UL))
```
Could you accept the above modification ?

cyliangtw replied:
According to @gmarull instructions, modified `numaker_m46x_clock.h/numaker_m46x_reset.h/numaker-m46x-pinctrl.h` to be namespaced as NUMAKER_....

Also rebase to resolve merge conflcit of `pinctrl/Kconfig & CMakeLists.txt`.
Please take a look.

cyliangtw replied:
@gmarull , @dcpleung ,

Finished the namespace of dt-bindings, please take another look.

cyliangtw replied:
@gmarull, 
not get your feedback over 3 weeks, 
is there any concern ?

cyliangtw committed:
board: arm: add support for nuvoton pfm m467

add support for nuvoton pfm m467 development board

Signed-off-by: cyliang tw <cyliang@nuvoton.com>

dcpleung reviewed (dismissed):
LGTM on UART.

gmarull reviewed (dismissed):
thanks for the continued effort!

gmarull commented:
please rebase

dcpleung reviewed (dismissed):
UART looks good.

MulinChao reviewed (dismissed):
LGTM :)

carlescufi commented:
@cyliangtw please rebase the PR

cyliangtw commented:
> @cyliangtw please rebase the PR

Rebase to resolve /serial/Kconfig merge conflict.
